### PR TITLE
feat(core): new LazyVim release v14.0

### PR DESCRIPTION
**Pull request by folke**
# What's new?

## 14.x

Big new release with a lot of changes and improvements!
Two new plugins have been added, and a lot of plugins have been replaced.
With these changes, default **LazyVim** is now just `34` plugins.

### Added Plugins

- [fzf-lua](https://github.com/ibhagwan/fzf-lua) as a replacement for [telescope.nvim](https://github.com/nvim-telescope/telescope.nvim)
  - to use **telescope.nvim** instead, enable the `ui.telescope` extra
- [blink.cmp](https://github.com/saghm/blink.cmp) as a replacement for [nvim-cmp](https://github.com/hrsh7th/nvim-cmp)
  - to use **nvim-cmp** instead, enable the `nvim-cmp` extra

### Removed Plugins

- [dressing.nvim](https://github.com/stevearc/dressing.nvim) (replaced with [fzf-lua](https://github.com/ibhagwan/fzf-lua) and [snacks.input](https://github.com/folke/snacks.nvim))
- [telescope-fzf-native.nvim](https://github.com/nvim-telescope/telescope-fzf-native.nvim) (replaced with [fzf-lua](https://github.com/ibhagwan/fzf-lua))
- [telescope.nvim](https://github.com/nvim-telescope/telescope.nvim) (replaced with [fzf-lua](https://github.com/ibhagwan/fzf-lua))
- [indent-blankline.nvim](https://github.com/lukas-reineke/indent-blankline.nvim) (replaced with [snacks.indent](https://github.com/Folke/snacks.nvim))
  - to use **indent-blankline.nvim** instead, enable the `indent-blankline` extra
- [nvim-cmp](https://github.com/hrsh7th/nvim-cmp) (replaced with [blink](https://github.com/Saghen/blink.cmp))
- [nvim-snippets](https://github.com/hrsh7th/vim-vsnip) (replaced with [blink](https://github.com/Saghen/blink.cmp))
- [cmp-buffer](https://github.com/hrsh7th/cmp-buffer) (replaced with [blink](https://github.com/Saghen/blink.cmp))
- [cmp-nvim-lsp](https://github.com/hrsh7th/cmp-nvim-lsp) (replaced with [blink](https://github.com/Saghen/blink.cmp))
- [cmp-path](https://github.com/hrsh7th/cmp-path) (replaced with [blink](https://github.com/Saghen/blink.cmp))

### Changes

- added [`snacks.input`](https://github.com/folke/snacks.nvim/blob/main/docs/input.md)
- added [`snacks.scroll`](https://github.com/folke/snacks.nvim/blob/main/docs/scroll.md)
- added [`snacks.indent`](https://github.com/folke/snacks.nvim/blob/main/docs/indent.md)
- added [`snacks.scope`](https://github.com/folke/snacks.nvim/blob/main/docs/scope.md)
- added [`snacks.dim`](https://github.com/folke/snacks.nvim/blob/main/docs/dim.md)
- added [`snacks.zen`](https://github.com/folke/snacks.nvim/blob/main/docs/zen.md)
- changed default [`which-key`](https://github.com/folke/which-key.nvim) preset to `helix`
- drop `LazyVim.ui.fg` in favor of [`Snacks.util.color`](https://github.com/folke/snacks.nvim/blob/main/docs/util.md)

To disable **all animations**, add the following to your `options.lua`:

```lua
vim.g.snacks_animate = false
```

### Keymaps

- `<leader>z` to toggle [zen mode](https://github.com/folke/snacks.nvim/blob/main/docs/zen.md)
- `<leader>Z` to toggle [zoom mode](https://github.com/folke/snacks.nvim/blob/main/docs/zen.md)
- `<leader>uD` to toggle [dimming](https://github.com/folke/snacks.nvim/blob/main/docs/dim.md)
- `<leader>ua` to toggle [animations](https://github.com/folke/snacks.nvim/blob/main/docs/animate.md)
- `<leader>uS` to toggle [scroll](https://github.com/folke/snacks.nvim/blob/main/docs/scroll.md)
- `<leader>ug` to toggle [indent guides](https://github.com/folke/snacks.nvim/blob/main/docs/indent.md)
- [`snacks.profiler`](https://github.com/folke/snacks.nvim/blob/main/docs/profiler.md) keymaps under `<leader>dp`

### CODE REVIEW

**Comment by dpetka2001** [comment]
`lang.git` needs `optional = true` for nvim-cmp.

**Comment by dpetka2001** [comment]
I've just started using this, but I'll be out for a couple of hours, so will continue testing later tonight and report back if I find anything else.

I also have these in my personal configuration
```lua
Snacks.toggle.indent():map("<leader>ug")
Snacks.toggle.scroll():map("<leader>uS")
```
Don't know if you want to include them by default, just thought I'd mention this since `indent` and `scroll` are enabled by default now.

**Comment by folke** [comment]
Added!

**Comment by dpetka2001** [comment]
Also just for clarification, credits go to @vdegenne for these mappings, since he mentioned them first in Snacks repo and I just borrowed them. I'm sure he'd make the same suggestion, I just happened to test this PR first.

**Comment by folke** [comment]
Any ideas for the profiler keymaps? I currently have this in my config:

```lua
  {
    "snacks.nvim",
    opts = function()
      Snacks.toggle.profiler():map("<leader>pp")
      Snacks.toggle.profiler_highlights():map("<leader>ph")
    end,
    -- stylua: ignore
    keys = {
      { "<leader>ps", function() Snacks.profiler.scratch() end },
    },
  },
```

That introduces a new global leader key, which is not ideal and also conflicts with `yanky`.

Other options:

- `<leader>dp`: good fit, but conflicts with `dap` pause. Not sure if that's used frequently and could be moved to `<leader>dP`
- `<leader>cp`: markdown preview at the moment.

Any better ideas? :)

**Comment by minusfive** [comment]
Wonder whether zen and zoom should be grouped under `w` (since they're window related), or `u` like the others (since they're ui related, and for consistency).

It'd be nice to keep the top level mappings as free as possible for users, and or other more frequently used functionality.

**Comment by folke** [comment]
Yep, good idea. Updated!

**Comment by folke** [comment]
Maybe move the profiling keymaps under `<leader>P`?
Since that's not something people will frequently use, unless they're developing plugins...

**Comment by folke** [comment]
Or go with the `dp` and move pause to `dP`

**Comment by folke** [comment]
Or remove dap pause and make dap continue to both?

**Comment by minusfive** [comment]
> Or go with the `dp` and move pause to `dP`

What about keeping `dp` as is and adding `dP` for the profiler?

**Comment by folke** [comment]
I don't think `dap` pause` is actually all that useful.
For the profiler there need to be keys under it, so a bit akward to type a lower-case key, then upper and lower again.
I'm leaning towards moving pasue to `dP`

**Comment by minusfive** [comment]
> I don't think `dap` pause is actually all that useful.
> 
> For the profiler there need to be keys under it, so a bit akward to type a lower-case key, then upper and lower again.
> 
> I'm leaning towards moving pasue to `dP`

Yeah that makes sense.

**Comment by minusfive** [comment]
While you're moving keymaps around and taking advantage of the breaking release, what do you think about moving the `Lazy` mappings to an `L` group, and leaving the top level `l` group for LSP-related functionality?

Examples:

- `<leader>lg` "Grammar" lsp (harper) toggle, since it's too noisy to leave on by default, but handy on demand 
- `<leader>Ll` Lazy
- `<leader>Ln` News
- `<leader>Lx` LazyExtras

**Comment by folke** [comment]
@saghen just a heads-up that I'm going to make `blink` the default completion engine for Neovim >= 0.10

File: lua/lazyvim/plugins/extras/ui/indent-blankline.lua [hunk 1]
```diff
@@ -0,0 +1,51 @@
+return {
+  -- disable snacks indent when animate is enabled
```
Comment (minusfive) [inline lua/lazyvim/plugins/extras/ui/indent-blankline.lua:2]: `s/animate/indent-blankline`

**Comment by dpetka2001** [comment]
Personally, not a fan of `<leader>l` for LSP related stuff. In my old configuration I had moved LSP related stuff under `<leader>cl`, since I think of them as part of `coding`. 

`<leader>l` for Lazy UI is so much useful as I use it the most and rarely use the others to justify the use of them in a group for Lazy stuff. 

I believe they should belong in a user personal configuration if someone wants, but that's just my personal opinion in the end.

**Comment by vdegenne** [comment]
Please do not change `<leader>l` !

**Comment by folke** [comment]
I agree with @dpetka2001, and also use `<leader>l` often

**Comment by vdegenne** [comment]
to restart lsp, `:lspr` then `<Tab><Return>` is not much for something you don't call often

**Comment by folke** [comment]
Overall, what do you think of all the changes? Most importantly the change to blink / fzf-lua etc.
I'm anticipating a lot of whining, but that's to be expected, since most people don't deal well with change.

On the other hand, I do think these changes are a better & cleaner set of plugins for a new LazyVim install.

File: lua/lazyvim/plugins/ui.lua [hunk 2]
```diff
@@ -308,7 +265,26 @@ return {
   { "MunifTanjim/nui.nvim", lazy = true },
 
   {
-    "folke/snacks.nvim",
+    "snacks.nvim",
```
Comment (minusfive) [inline lua/lazyvim/plugins/ui.lua:268]: Noticed you have these as `"folke/snacks.nvim"` on `init.lua`, but dropped the `"folke"` here.

(P.S. how does that work?)

Comment (folke) [reply lua/lazyvim/plugins/ui.lua:268]: Yes, that works. As long as it's specified in full somewhere.

Comment (minusfive) [reply lua/lazyvim/plugins/ui.lua:268]: TIL! Thanks!

**Comment by dpetka2001** [comment]
I've been using `fzf-lua` since it first appeared as an Extra, so really in favor of this decision. And I recently re-enabled `blink.cmp` as I found a way to make it not conflict with `vim-visual-multi` and break accepting suggestions. 

Just my personal opinion, but really in favor of these defaults as a simple user.

**Comment by minusfive** [comment]
I'm a fan so far. Yeah, people will complain whenever there's change 🤷.

I'll be testing this all day and report back.

**Comment by dpetka2001** [comment]
And `:LazyExtras` makes it really easy to enable/disable Extras according to what a user wants to use. It really shouldn't be so much of a problem for users to change to Telescope or nvim-cmp.

**Comment by folke** [comment]
Agree.
Animated scroll is also part of this release.
I did add a `vim.g.snacks_animate = true` that you can set to `false` to disable **all** snacks animations, including scroll, indent, ...

**Comment by folke** [comment]
My work is having fun working on Neovim plugins, so yes I might be a bit biased 😅.

But even when I'm messing with AI projects instead, I'm still using Lazy all the time...

**Comment by folke** [comment]
Fun fact: I mainly create plugins to improve my coding experience while creating plugins 😅

So every now and then I can just dissappear for a couple of months. If I'm not working on plugins, then I literally have no need to have better plugins for Neovim, so no need to work on plugins.

I'm a walking self-fulfilling prophecy 😂

**Comment by folke** [comment]
Merged :)

Will wait at least till tonight to make an actual new release, so that people that are following `main` can already share their input (and bugs).

**Comment by minusfive** [comment]
> Fun fact: I mainly create plugins to improve my coding experience while creating plugins 😅
> 
> 
> 
> So every now and then I can just dissappear for a couple of months. If I'm not working on plugins, then I literally have no need to have better plugins for Neovim, so no need to work on plugins.
> 
> 
> 
> I'm a walking self-fulfilling prophecy 😂

I'd argue it's in the interest of the community to keep you happy and productive man, so please remap `<esc>` to Lazy if you need to! 🤣

**Comment by Krystian-KF** [comment]
Looks like clangd extension is stucked with nvim-cmp. 
Screenshot_20241212_143558

**Comment by folke** [comment]
@Krystian-KF should be fixed now

**Comment by skoch13** [comment]
I tried `blink` prior to this merge and it was significantly slower with TS in my tests, yet I didn't have time to investigate why.

**Comment by Krystian-KF** [comment]
There is small discrepancy between News.md and the config:
- toggle [zen mode] is set to `<leader>uz`, whereas News.md says it is `<leader>z`
- toggle [zoom mode] is set to `<leader>uZ`, whereas News.md says it is `<leader>Z`

One quick question if you don't mind. 
How close is Snacks's zen mode to zen-mode.nvim, does it accept `on_open` and `on_close` callbacks or it is better to stay with the plugin to keep using those? Sorry for asking but haven't had a chance yet to check it myself.

**Comment by folke** [comment]
@Krystian-KF just added the `on_open`, `on_close` callbacks. Check the updated docs.
Also fixed the `NEWS.md`.

As for other functionality, I'm probably going to create an `emulator` snacks, to have a general interface to talk to your terminal emulator (whether it's wezterm, kitty, ...). You'll then also be able to use that functionality in zen mode.

**Comment by sibouras** [comment]
on windows fzf-lua is noticeably slow because the fzf process takes 55ms to start on windows compared to 4ms on linux(wsl)

**Comment by dpetka2001** [comment]
So, just use Telescope on Windows? You can still enable it from `:LazyExtras`. Each one to his own preferences.

**Comment by sibouras** [comment]
i know that, but making it default is not good for new users. they will think lazyvim is slow

**Comment by dpetka2001** [comment]
Then they should ask a question and they will get an answer if they describe what they observe. Either way maintainer has the final say and I'm sure he'll see your comment and make his decision. I don't mind either way, since I was using `fzf-lua` even when it wasn't default.

**Comment by folke** [comment]
@sibouras I've been using fzf lua on windows for testing and never experienced that delay you're talking about, so there must be something off with your config...

**Comment by sibouras** [comment]
in this gif telescope is snappy when opening and closing and fzf-lua is noticeably slower, also when going down with `j` fzf-lua is also noticeably slower.

 im using ryzen 5 5500u, ur probably using a beefy machine if you don't notice anything. if im the only one noticing this im fine with fzf-lua as default

https://github.com/user-attachments/assets/21e48dc0-e6d0-4f5a-9c73-b73cf62b0f25

**Comment by TheJoeSchr** [comment]
Link to blink is borked in announcement, correct link is: https://github.com/Saghen/blink.cmp

**Comment by folke** [comment]
That has already been fixed, unless you see it wrong somewhere else as well?

**Comment by TheJoeSchr** [comment]
> That has already been fixed, unless you see it wrong somewhere else as well?

I see it wrong at the top of the PR:
image

It links to: https://github.com/saghm/blink.cmp (404)

**Comment by folke** [comment]
ok, but this PR is closed, so who cares.
I thought you were referring to the NEWS.md

**Comment by TheJoeSchr** [comment]
> ok, but this PR is closed, so who cares. I thought you were referring to the NEWS.md

Ok, I just came here after reading the news in nvim to seek out the reasoning behind the telescope change and then came across the broken link. Thought to safe others some troubles

**Comment by Daydreamer-riri** [comment]
> in this gif telescope is snappy when opening and closing and fzf-lua is noticeably slower, also when going down with `j` fzf-lua is also noticeably slower.
> 
> im using ryzen 5 5500u, ur probably using a beefy machine if you don't notice anything. if im the only one noticing this im fine with fzf-lua as default
> 
>  fzf-vs-telescope.mp4

fzf-lua is slow on wsl too.

**Comment by folke** [comment]
It's faster to enable telescope with `:LazyExtras` than it is to complain here. smh

**Comment by Daydreamer-riri** [comment]
> It's faster to enable telescope with `:LazyExtras` than it is to complain here. smh

Sorry, I didn't want to complain, it was my expression error. I'd love to use it, and I've just been trying to figure out why it's slow to start. I have some strange things. Even if I am on wsl, fzf-lua is still about 100 milliseconds slower than telescope (feeling).

**Comment by vdegenne** [comment]
btw, for now `blink.cmp` seems to work better with LuaSnip `2.3.0` but LazyVim is using `main` branch by default.

**Comment by dpetka2001** [comment]
@vdegenne LazyVim is using blink latest stable release by default. Only if you put `vim.g.lazyvim_blink_main = true` in your `options.lua` you will use blink's main branch.

**Comment by danrasmuson** [comment]
I also came to this thread wondering why fzf is the recommended default rather then telescope. My understanding is that a team has worked hard to custom build telescope to integrate effectively with neovim and fzf is a terminal based tool that does have the same quality of integration with neovim. I'm happy to opt-into staying with telescope, but if I'm missing something about the value add of fzf as the default I would love to know.

**Comment by danrasmuson** [comment]
> > It's faster to enable telescope with `:LazyExtras` than it is to complain here. smh
> 
> Sorry, I didn't want to complain, it was my expression error. I'd love to use it, and I've just been trying to figure out why it's slow to start. I have some strange things. Even if I am on wsl, fzf-lua is still about 100 milliseconds slower than telescope (feeling).

I also came to this thread wondering why fzf is the recommended default rather then telescope. My understanding is that a team has worked hard to custom build telescope to integrate effectively with neovim and fzf is a terminal based tool that does have the same quality of integration with neovim. I'm happy to opt-into staying with telescope, but if I'm missing something about the value add of fzf as the default I would love to know.

**Comment by vdegenne** [comment]
@dpetka2001 Yes I meant the main branch of LuaSnip seems to be unstable for blink (main or stable). Tab works a little bit better but there are still some quirks.

**Comment by dpetka2001** [comment]
@vdegenne You can do the following if you want
```lua
{
    "L3MON4D3/LuaSnip",
    version = "*",
  },
```
This will force `lazy.nvim` to install the latest stable version of Luasnip. Except if there are new features not available in the latest stable release and you want to use main branch explicitly.

**Comment by vdegenne** [comment]
Yes I'm already doing that, I was just making a suggestion, if the main branch of LuaSnip causes some issues maybe LazyVim should temporarily avoid it. If new users try LazyVim for the first time but snippets don't work as intended...
(Not complaining personally.)

**Comment by folke** [comment]
New users should not use luasnip. Native snippets are more than enough.

**Comment by xzbdmw** [comment]
> I also came to this thread wondering why fzf is the recommended default rather then telescope. My understanding is that the telescope team has worked hard to custom build telescope effectively into neovim, while fzf is a terminal based tool that doesn't have the same quality of integration with neovim. I'm happy to opt-into staying with telescope, but if I'm missing something about the value add of fzf as the default I would love to know.

To my understanding,
pros: speed, much faster in large repos (goroutins vs single thread lua coroutines);
cons: no normal mode keymaps.

**Comment by vdegenne** [comment]
> New users should not use luasnip. Native snippets are more than enough.

My mistake, I thought blink used LuaSnip as the default but it was just registered somewhere in my specs, now that I removed it, it works better.
Btw, if I `<leader>l` to see loaded modules I don't see `LuaSnip` anymore as expected, but I don't see `nvim-snippets` either. What does blink use as the snippet engine? Unless blink itself also serve as a snippet engine?

**Comment by folke** [comment]
nvim-snippets is also not needed with blink.

**Comment by folke** [comment]
Neovim has a native snippet engine and blink covers what's additionally done with nvim-snippets

**Comment by vdegenne** [comment]
Great to know Thx

**Comment by folke** [comment]
Was to be expected 😅
https://www.reddit.com/r/neovim/comments/1he11s0/lazy_constantly_replacing_plugins_and_breaking/

**Comment by dpetka2001** [comment]
looool

**Comment by vdegenne** [comment]
on first install you should pop a notice with two choices, one to agree to changes or not in which case it would just kill the install :roll_eyes:

**Comment by dpetka2001** [comment]
People should just get comfortable with the tools they use. That's the whole meaning of using Neovim. But to get there, it means that you will have to read some docs about how things are done. LazyVim is not a LTS Linux distro. It's more like rolling release one. 

And honestly `:LazyExtras` are one of the best things since the birth of this distro. I haven't seen this in other distros (although it has been quite a bit of time since I last checked them, so maybe things have changed in the mean time). If other distros make a breaking change how do users cope with it? I'm sure it would require more effort. With `LazyExtras` you just switch to the Extra you want and make sure you don't have any unnecessary plugin spec in your personal configuration to avoid breakages. I really don't know how much easier than that can a breaking transition be.

**Comment by folke** [comment]
I think a lot of users, don't read breaking changes in the Lazy ui and just close the news when it opens...

**Comment by vdegenne** [comment]
@folke I have a new suggestion, how hard would it be to allow this syntax?

While this works:

```lua
Snacks.config.style("news", {
  width = 0.6,
  height = 0.6,
  wo = {
    spell = false,
    wrap = false,
    signcolumn = "yes",
    statuscolumn = " ",
    conceallevel = 3,
  },
})

local float = Snacks.win({
  relative = "editor",
  file = file,
  style = "news",
})
```

This should also works:

```lua
local style = Snacks.config.style("news", {
  width = 0.6,
  height = 0.6,
  wo = {
    spell = false,
    wrap = false,
    signcolumn = "yes",
    statuscolumn = " ",
    conceallevel = 3,
  },
})

local float = Snacks.win({
  relative = "editor",
  file = file,
  style = local,
})
```

That would help for reading code in general because using imports we can easily use `gd` to reach the definition of a style. Sometimes the styles are declared far from the window definition, which makes code learning cumbersome.

**Comment by vdegenne** [comment]
I guess that wouldn't be so hard to implement, `Snacks.config.style` can just return the name (`news` in this example). I can make the PR if you agree with this.

**Comment by Parsifa1** [comment]
> in this gif telescope is snappy when opening and closing and fzf-lua is noticeably slower, also when going down with `j` fzf-lua is also noticeably slower.
> 
> im using ryzen 5 5500u, ur probably using a beefy machine if you don't notice anything. if im the only one noticing this im fine with fzf-lua as default

This may be a limitation of windows :(
ref https://github.com/ibhagwan/fzf-lua/discussions/1353

**Pull request merged**